### ISSUE TITLE
RDataFrame and CreatePoxy/TTreeReaderValue error for selection on string branch in Filter method

### ISSUE THREAD

**Issue by romainbqt** [root-project/root#7169]
Hi, 
I create this issue to report a bug, that was also observed by @eguiraud, 
with ROOT version 6.22.06

I am trying to use the RDataFrame to produce histograms running on root files.
With the following code
```python
import ROOT as RT 
"""
# setupATLAS and ROOT 6.22.06 
export ATLAS_LOCAL_ROOT_BASE=/cvmfs/atlas.cern.ch/repo/ATLASLocalRootBase
alias setupATLAS='source ${ATLAS_LOCAL_ROOT_BASE}/user/atlasLocalSetup.sh'
setupATLAS
. /cvmfs/sft.cern.ch/lcg/app/releases/ROOT/6.22.06/x86_64-centos7-gcc48-opt/bin/thisroot.sh

# Execute python script
python2 dataframe_vs_tree.py
"""
RT.gROOT.SetBatch(True) 
RT.EnableImplicitMT()

infileName    = "Zee_Sh221-0.root" 
treeName      = "Nominal"  
cutString     = "sample==\"Zbl\"" 
distName      = "MET" 
#weightName    = "EventWeight" 
binning       = [25,0,500]

histName = "h_{}".format(distName)

print("infileName = {}".format(infileName))
print("treeName   = {}".format(treeName))
print("cutString  = {}".format(cutString)) 

infile = RT.TFile.Open(infileName)
dataframe = RT.RDataFrame(treeName, infile)

histmodel = RT.RDF.TH1DModel(histName, histName, binning[0], binning[1], binning[2])

hist = dataframe.Filter(cutString).Histo1D(histmodel, distName)  
print("-----")
print("hist.GetEntries() = {}".format(hist.GetEntries()))
print("hist.GetSumOfWeights()= {}".format(hist.GetSumOfWeights()))
print("-----")
```
I attached the test file on the ROOT forum (which only contains 3 entries on purpose)
Please read the following strange observations that we made about it which to us do not make sense 
https://root-forum.cern.ch/t/rdataframe-and-createpoxy-ttreereadervalue-error-for-selection-on-string-branch-in-filter-method/43488

But I get the error
```
Error in <TTreeReaderValueBase::CreateProxy()>: The branch sample contains data of type string. It cannot be accessed by a TTreeReaderValue<string>
```
With bigger files I get many times this error and the number of entries and integral of the histogram hist is 0 while it should not be . Since for the current file doing a scan of the tree I have:
```
root -l Zee_Sh221-0.root 
root [0] 
Attaching file Zee_Sh221-0.root as _file0...
(TFile *) 0x4418460
root [1] Nominal->Scan("sample:MET")
************************************
*    Row   *    sample *       MET *
************************************
*        0 *        Zl * 158.14303 *
*        1 *       Zbl * 180.14964 *
*        2 *       Zbl * 253.89321 *
************************************
```
This error only occurs when the selection in Filter function includes a branch of type string.
If I have a selection only on double, float and so on without string branch everything is fine.

Would you know how to solve that ?
I also tried using python3 but same error occurs.
And using the corresponding TTree command works so it does not comes from the root file:
```
root -l Zee_Sh221-0.root
root [0] 
Attaching file Zee_Sh221-0.root as _file0...
root [1] TTree *tree = (TTree*) _file0->Get("Nominal") 
root [2] tree->Draw("MET>>h_MET(25,0,500)", "sample==\"Zbl\"")
Info in <TCanvas::MakeDefCanvas>:  created default TCanvas with name c1
(long long) 2
```

**Comment by etejedor** [comment]
A simpler reproducer, generating a tree with:

```cpp
void generate_tree() {
  TFile f("file.root", "RECREATE");
  TTree t("tree", "tree");
  std::string x("foo");
  t.Branch("x", &x);
  t.Fill();
  f.Write();
  f.Close();
}
```

and then running:
```python
# tree generated in C++
import ROOT

df = ROOT.RDataFrame('tree', 'file.root')
hist = df.Histo1D('x')

ROOT.std.string

print(hist.GetEntries())
```

triggers the same error. The culprit here is the lookup of `std::string`.

**Comment by etejedor** [comment]
A simpler reproducer, after digging what triggers the problem in the cppyy code:

```python
# tree generated in C++
import ROOT

df = ROOT.RDataFrame('tree', 'file.root')
hist = df.Histo1D('x')

ROOT.gROOT.GetType("std::string") # this triggers the issue

print(hist.GetEntries())
```

So it does not seem a PyROOT problem. @Axel-Naumann and perhaps @vgvassilev any comments on this?

**Comment by etejedor** [comment]
A C++ reproducer:

```cpp
void repro() {
  auto df = ROOT::RDataFrame("tree", "file.root");
  auto hist = df.Histo1D("x");

  gROOT->GetType("std::string");

  cout << hist->GetEntries() << endl;
}
```

**Comment by eguiraud** [comment]
(use `Histo1D<std:string>` to get everything compiled)

**Comment by eguiraud** [comment]
A compilable TTreeReader-only repro:

```cpp
#include <TTreeReader.h>
#include <TTreeReaderValue.h>
#include <TFile.h>
#include <TTree.h>
#include <TROOT.h>
#include <string>
#include <iostream>

void MakeTree()
{
   TFile f("f.root", "recreate");
   TTree t("t", "t");
   std::string s("foo");
   t.Branch("s", &s);
   t.Fill();
   f.Write();
   f.Close();
}

int main() {
   MakeTree();

   TFile f("f.root");
   TTreeReader r("t", &f);
   TTreeReaderValue<std::string> rv(r, "s");

   gROOT->GetType("std::string"); // this is what breaks it

   while (r.Next())
      std::cout << *rv << '\n';

   return 0;
}
```

This prints:

```
Error in <TTreeReaderValueBase::CreateProxy()>: The branch s contains data of type string. It cannot be accessed by a TTreeReaderValue<string>
```

The difference between runs with/without the `GetType("std::string")` seems to be that at the following line `fDict != branchActualType` when `GetType` is called, while `fDict == branchActualType` in the sane case. I did not have time to go deeper.

https://github.com/root-project/root/blob/master/tree/treeplayer/src/TTreeReaderValue.cxx#L557

**Comment by eguiraud** [comment]
@romainbqt did you ever manage to work around the issue? otherwise i'll try to take a look as soon as possible

**Comment by romainbqt** [comment]
Hi @eguiraud,    
No I did not tried further to find a solution, this issue was to report that bug.   
It would be nice for it to be fixed since using strings is pretty common    
Thanks again for having a look into that problem

**Comment by eguiraud** [comment]
Ok, I'll do my best to take a deeper look this week :+1:

**Comment by eguiraud** [comment]
Thanks to @Axel-Naumann this is now fixed in ROOT's master branch. Do we want to backport to 6.24?

**Comment by romainbqt** [comment]
Thanks @Axel-Naumann and @eguiraud for the fix !

**Comment by eguiraud** [comment]
The fix was backported to 6.24 patches (will be available in 6.24/02), closing this.
